CMakeLists.txt

```diff
@@ -13,22 +14,63 @@ project(
 
 include(GNUInstallDirs)
 
+option(BUILD_TESTING "Top-level switch for testing. Turn off to disable unit and contract tests." ON)
+
+option(LD_BUILD_SHARED_LIBS "Build the SDKs as shared libraries" OFF)
+
+cmake_dependent_option(LD_BUILD_UNIT_TESTS
+        "Build the C++ unit tests."
+        ON                                      # default to enabling unit tests
+        BUILD_TESTING;NOT LD_BUILD_SHARED_LIBS  # only exposed if top-level switch is on, and also only when building
+        # static libs. This is because we have hidden visibility of symbols by default (to only expose our C API.)
+        OFF                                     # otherwise, off
+)
+
+# If you want to run the unit tests with valgrind, then LD_TESTING_SANITIZERS must of OFF.
+cmake_dependent_option(LD_TESTING_SANITIZERS
+        "Enable sanitizers for unit tests."
+        ON                      # default to enabling sanitizers
+        LD_BUILD_UNIT_TESTS        # only expose if unit tests enabled..
+        OFF                     # otherwise, off
+)
+
+cmake_dependent_option(LD_BUILD_CONTRACT_TESTS
+        "Build contract test service."
+        OFF             # default to disabling contract tests, since they require running a service
+        BUILD_TESTING   # only expose if top-level switch is on..
+        OFF             # otherwise, off
+)
+
+# The general strategy is to produce a fat artifact containing all of our dependencies so users
+# only have a single thing to link. We should support this either being a static or shared library.
+# Because OpenSSL is a large, and security relevant dependency, we should have a separate option
+# to link against that statically or dynamically.
+
+option(LD_STATIC_LINK_OPENSSL
+        "Link OpenSSL as a static library or as a shared library"
+        ON  # default to linking OpenSSL statically
+)
+
+option(LD_BUILD_EXAMPLES "Build hello-world examples." ON)
+
+
```

Comment:
Throwing a warning when someone tries to build shared lib with unit tests enabled. This won't work since we hide symbols by default.